py/visdom/server.py

```diff
@@ -428,6 +432,142 @@ def on_close(self):
+            json.dumps({'command': 'register', 'data': self.sid,
+                        'readonly': self.readonly}))
+        self.broadcast_layouts([self])
+        broadcast_envs(self, [self])
+
+    def on_message(self, message):
+        logging.info('from web client: {}'.format(message))
+        msg = tornado.escape.json_decode(tornado.escape.to_basestring(message))
+
+        cmd = msg.get('cmd')
+
+        if self.readonly:
+            return
+
+        if cmd == 'close':
+            if 'data' in msg and 'eid' in msg:
+                logging.info('closing window {}'.format(msg['data']))
+                p_data = self.state[msg['eid']]['jsons'].pop(msg['data'], None)
+                event = {
+                    'event_type': 'Close',
+                    'target': msg['data'],
+                    'eid': msg['eid'],
+                    'pane_data': p_data,
+                }
+                send_to_sources(self, event)
+        elif cmd == 'save':
+            # save localStorage window metadata
+            if 'data' in msg and 'eid' in msg:
+                msg['eid'] = escape_eid(msg['eid'])
+                self.state[msg['eid']] = \
+                    copy.deepcopy(self.state[msg['prev_eid']])
+                self.state[msg['eid']]['reload'] = msg['data']
+                self.eid = msg['eid']
+                serialize_env(self.state, [self.eid], env_path=self.env_path)
+        elif cmd == 'delete_env':
+            if 'eid' in msg:
+                logging.info('closing environment {}'.format(msg['eid']))
+                del self.state[msg['eid']]
+                if self.env_path is not None:
+                    p = os.path.join(
+                        self.env_path,
+                        "{0}.json".format(msg['eid'])
+                    )
+                    os.remove(p)
+                broadcast_envs(self)
+        elif cmd == 'save_layouts':
+            if 'data' in msg:
+                self.app.layouts = msg.get('data')
+                self.app.save_layouts()
+                self.broadcast_layouts()
+        elif cmd == 'forward_to_vis':
+            packet = msg.get('data')
+            environment = self.state[packet['eid']]
+            packet['pane_data'] = environment['jsons'][packet['target']]
+            send_to_sources(self, msg.get('data'))
+        elif cmd == 'layout_item_update':
+            eid = msg.get('eid')
+            win = msg.get('win')
+            self.state[eid]['reload'][win] = msg.get('data')
+
+    def close(self):
+        if self in list(self.subs.values()):
+            self.subs.pop(self.sid, None)
+
+    def write_message(self, msg):
+        self.messages.append(msg)
+
+    def get_messages(self):
+        to_send = []
+        while len(self.messages) > 0:
+            message = self.messages.pop()
+            if type(message) is dict:
+                # Not all messages are being formatted the same way (JSON)
+                # TODO investigate
```

Comment:
what is the symptom of this, i.e. in what unintended way are these messages being formatted? what are potential implications?

Reply:
The symptom was that sometimes the packet sent to the frontend via polling would not be properly encoded for sending, and would fail to be decoded on the other end. This is actually super inefficient to be Json encoding windows, so if it turns out to be possible to not encode them individually I want to swap to doing that. For some reason the socket script was doing fine at decoding both cases, which I didn't take the time to fully understand during this pass

Reply:
hmm okay, if it's not causing any major issues then it's fine by me.

Reply:
Yeah if anything this just happened to expose that JSON encoding magic has already been going on behind the scenes and I should examine that at some point.